### PR TITLE
File browser interaction cleanup

### LINKED ISSUE #441: Right clicking on filebrowser item should activate the item
In addition to opening the context menu, right clicking on an item in the file browser should *also* activate the item, as if it were left clicked.

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #441.

Context Menus:
- Right clicking on an item selects it before opening the context menu
- Right clicking on an unselected item when a context menu is open selects a new item and opens the context menu
- A left click anywhere when there is a context menu clears the menu but keeps the selection

Soft Selection:
- Unmodified mouse down an item in an existing selection is a "soft selection" for dragging or context menu
- Mouse up on a soft selection clears other selection unless there is an active context menu

Drag-drop:
- Fixes a bug when dragging multiple items - the original node text was being changed
- Uses a file icon for all multi-item drags

There is overall less spurious triggering of file name editing as well.

fix-file-selection

### CODE REVIEW

File: src/filebrowser/listing.ts [hunk 1]
```diff
@@ -702,6 +686,13 @@ class DirListing extends Widget {
   }
 
   /**
+   * Handle the `'contextmenu'` event for the widget.
+   */
+  private _evtContextMenu(event: MouseEvent): void {
+    this._inContext = true;
```
Comment (sccolbert) [inline src/filebrowser/listing.ts]: IIRC `contextmenu` events are fired *after* `mousedown` events

Comment (blink1073) [inline src/filebrowser/listing.ts]: That's fine, I need it for the *next* mousedown event.

File: src/filebrowser/listing.ts [hunk 2]
```diff
@@ -354,14 +354,10 @@ class DirListing extends Widget {
...
-        if (this._selection[item.name]) {
-          names.push(item.name);
-        }
+    let items = this._model.items;
+    for (let item of items) {
+      if (this._selection[item.name]) {
+        names.push(item.name);
       }
     }
     let message = `Permanantly delete these ${names.length} files?`;
```
Comment (afshin) [inline src/filebrowser/listing.ts]: `Permanantly` => `Permanently`

This message may be worded better altogether, I'm not sure.

Comment (blink1073) [inline src/filebrowser/listing.ts]: Spelling fixed.

**Comment by jasongrout** [comment]
> Right clicking on an unselected item when a context menu is open selects a new item and opens the context menu

There are two ways to do right-click on mac. A two-finger click, which works like you want (i.e., it clears the old selection, selects the new item, and opens the context menu). However, ctrl+click is also the standard way to do right click, and that way selects the new item and *doesn't* clear the old selection.

**Comment by blink1073** [comment]
I'm having a hard time believing that the interacting in my Windows VM are truly representative.  I'll log on to my Windows desktop later today.  My preference would be to use the Windows behavior throughout, so as to be consistent at least.

**Comment by jasongrout** [comment]
Yep, +1 to the windows selection behavior. I'm just pointing out that something seems off in the click modifier checking on mac, since ctrl+click isn't the same as right click.

**Comment by jasongrout** [comment]
To be clearer, here's a screencast. I'm right clicking when the cursor is on the right side of the filebrowser. The first time, I'm right clicking several times using ctrl+click. The second time I'm right clicking using the two-finger click. All on OS X Chrome.

clicking

**Comment by blink1073** [comment]
The behavior on Windows is that Ctrl+Click is a toggle.
If we want to be purely consistent with Windows behavior, we could make Cmd+Click do nothing on Mac.
I think that is a fair thing to do, and leave any Mac-specific interactions as an enhancement that we may add later on.

**Comment by jasongrout** [comment]
+1 on having windows *behavior* in the select box, but -1 on insisting on having the same way to invoke it as on Windows. I think on a mac, Command-click should be the thing to invoke toggle, and ctrl-click should be passed through (it's already interpreted as right click).

**Comment by blink1073** [comment]
I'll try that out and we'll see how it feels.

**Comment by jasongrout** [comment]
Thanks!

**Comment by blink1073** [comment]
@jasongrout, updated so that Cmd+Click toggles on Mac, and Ctrl+Click behaves the same as right click on Mac.

**Comment by jasongrout** [comment]
Works great; thanks!

As a bonus, if I do Ctrl+Cmd+click, I get to simulaneously open the context menu and toggle a new item!

**Pull request merged**